File: protocol/runtime/go/contract_test.go
package protocolruntime

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type fixtureSet struct {
	SchemaCases []struct {
		Name       string          `json:"name"`
		Definition string          `json:"definition"`
		Valid      bool            `json:"valid"`
		Value      json.RawMessage `json:"value"`
	} `json:"schemaCases"`
	CanonicalCases []struct {
		Name      string          `json:"name"`
		Domain    string          `json:"domain"`
		Value     json.RawMessage `json:"value"`
		Canonical string          `json:"canonical"`
		SHA256    string          `json:"sha256"`
	} `json:"canonicalCases"`
	CanonicalRejectCases []struct {
		Name  string          `json:"name"`
		Value json.RawMessage `json:"value"`
	} `json:"canonicalRejectCases"`
}

func loadContractFixtures(t *testing.T) (*Validator, fixtureSet) {
	t.Helper()
	schemaRaw, err := os.ReadFile(filepath.Join("..", "..", "schemas", "tdev.v1.schema.json"))
	if err != nil {
		t.Fatal(err)
	}
	validator, err := ParseSchema(schemaRaw)
	if err != nil {
		t.Fatal(err)
	}
	fixtureRaw, err := os.ReadFile(filepath.Join("..", "..", "testdata", "fixtures.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fixtures fixtureSet
	if err := json.Unmarshal(fixtureRaw, &fixtures); err != nil {
		t.Fatal(err)
	}
	return validator, fixtures
}

func TestRawJSONAdmissionRejectsMalformedInput(t *testing.T) {
	t.Run("trailing value", func(t *testing.T) {
		if _, err := ParseJSON([]byte(`{} {}`)); err == nil {
			t.Fatal("expected trailing JSON to be rejected")
		}
	})
	t.Run("invalid UTF-8 value", func(t *testing.T) {
		if _, err := ParseJSON([]byte{'{', '"', 'x', '"', ':', '"', 0xff, '"', '}'}); err == nil {
			t.Fatal("expected invalid UTF-8 protocol JSON to be rejected")
		}
	})
	t.Run("invalid UTF-8 schema", func(t *testing.T) {
		if _, err := ParseSchema([]byte{'{', '"', '$', 'd', 'e', 'f', 's', '"', ':', '{', '"', 0xff, '"', ':', '{', '}', '}', '}'}); err == nil {
			t.Fatal("expected invalid UTF-8 schema JSON to be rejected")
		}
	})
}

func TestSchemaSubsetFailsClosedAndSafeNumberProfile(t *testing.T) {
	invalidSchemas := map[string]string{
		"unsupported keyword": `{"$defs":{"Probe":{"type":"string","unsupportedKeyword":true}}}`,
		"ref sibling":         `{"$defs":{"Probe":{"$ref":"#/$defs/Target","type":"string"},"Target":{"type":"string"}}}`,
		"oneOf sibling":       `{"$defs":{"Probe":{"oneOf":[{"type":"string"}],"type":"string"}}}`,
		"root keyword":        `{"$defs":{"Probe":{"type":"string"}},"allOf":[{"$ref":"#/$defs/Probe"}]}`,
		"malformed required":  `{"$defs":{"Probe":{"type":"object","properties":{},"required":"x","additionalProperties":false}}}`,
		"malformed minimum":   `{"$defs":{"Probe":{"type":"integer","minimum":"1"}}}`,
		"malformed enum":      `{"$defs":{"Probe":{"enum":"x"}}}`,
		"misplaced pattern":   `{"$defs":{"Probe":{"type":"integer","pattern":"x"}}}`,
		"misplaced format":    `{"$defs":{"Probe":{"type":"integer","format":"date-time"}}}`,
		"unsupported format":  `{"$defs":{"Probe":{"type":"string","format":"email"}}}`,
		"self alias cycle":    `{"$defs":{"Probe":{"$ref":"#/$defs/Probe"}}}`,
		"mutual alias cycle":  `{"$defs":{"A":{"$ref":"#/$defs/B"},"B":{"$ref":"#/$defs/A"}}}`,
		"oneOf self cycle":    `{"$defs":{"Probe":{"oneOf":[{"type":"string"},{"$ref":"#/$defs/Probe"}]}}}`,
		"trailing schema":     `{"$defs":{"Probe":{"type":"string"}}} {}`,
	}
	for name, raw := range invalidSchemas {
		t.Run(name, func(t *testing.T) {
			if _, err := ParseSchema([]byte(raw)); err == nil {
				t.Fatal("expected invalid schema to be rejected")
			}
		})
	}

	validator, err := ParseSchema([]byte(`{"$defs":{"NumberProbe":{"type":"number","minimum":1,"maximum":3}}}`))
	if err != nil {
		t.Fatal(err)
	}
	for raw, valid := range map[string]bool{"2": true, "0": false, "4": false, "1.5": false} {
		value, err := ParseJSON([]byte(raw))
		if err != nil {
			t.Fatal(err)
		}
		if got := len(validator.ValidateDefinition("NumberProbe", value)) == 0; got != valid {
			t.Fatalf("value %s valid=%v want=%v", raw, got, valid)
		}
	}
}

func TestContractFixturesAndCanonicalVectors(t *testing.T) {
	validator, fixtures := loadContractFixtures(t)
	semantic := map[string]bool{"NewCaseContractInput": true, "CaseContract": true, "NewCaseTargetGrant": true, "CaseTargetGrant": true}
	for _, fixture := range fixtures.SchemaCases {
		t.Run("schema/"+fixture.Name, func(t *testing.T) {
			value, err := ParseJSON(fixture.Value)
			if err != nil {
				t.Fatal(err)
			}
			var errors []string
			if semantic[fixture.Definition] {
				errors = ValidateContract(validator, fixture.Definition, value)
			} else {
				errors = validator.ValidateDefinition(fixture.Definition, value)
			}
			if got := len(errors) == 0; got != fixture.Valid {
				t.Fatalf("valid=%v errors=%v", got, errors)
			}
		})
	}
	for _, fixture := range fixtures.CanonicalCases {
		t.Run("canonical/"+fixture.Name, func(t *testing.T) {
			value, err := ParseJSON(fixture.Value)
			if err != nil {
				t.Fatal(err)
			}
			canonical, err := Canonicalize(value)
			if err != nil {
				t.Fatal(err)
			}
			if string(canonical) != fixture.Canonical {
				t.Fatalf("canonical=%s want=%s", canonical, fixture.Canonical)
			}
			digest, err := TypedDigest(fixture.Domain, value)
			if err != nil {
				t.Fatal(err)
			}
			if digest != fixture.SHA256 {
				t.Fatalf("digest=%s want=%s", digest, fixture.SHA256)
			}
		})
	}
	for _, fixture := range fixtures.CanonicalRejectCases {
		t.Run("reject/"+fixture.Name, func(t *testing.T) {
			value, err := ParseJSON(fixture.Value)
			if err != nil {
				return
			}
			if _, err := Canonicalize(value); err == nil {
				t.Fatal("expected parsing or canonicalization failure")
			}
		})
	}
}

File: protocol/contract.test.ts
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";
import type { CaseContract } from "./generated/typescript/types.ts";
import { canonicalize, typedDigest } from "./runtime/typescript/canonical.ts";
import { SchemaValidator, validateContract, type SchemaDocument } from "./runtime/typescript/schema.ts";

type FixtureSet = {
  schemaCases: { name: string; definition: string; valid: boolean; value: unknown }[];
  canonicalCases: { name: string; domain: string; value: unknown; canonical: string; sha256: string }[];
  canonicalRejectCases: { name: string; value: unknown }[];
};

const schema = JSON.parse(await readFile(new URL("./schemas/tdev.v1.schema.json", import.meta.url), "utf8")) as SchemaDocument;
const fixtures = JSON.parse(await readFile(new URL("./testdata/fixtures.json", import.meta.url), "utf8")) as FixtureSet;
const validator = new SchemaValidator(schema);
const semanticDefinitions = new Set(["NewCaseContractInput", "CaseContract", "NewCaseTargetGrant", "CaseTargetGrant"]);

test("schema and semantic fixtures", () => {
  for (const fixture of fixtures.schemaCases) {
    const errors = semanticDefinitions.has(fixture.definition)
      ? validateContract(validator, fixture.definition, fixture.value)
      : validator.validateDefinition(fixture.definition, fixture.value);
    assert.equal(errors.length === 0, fixture.valid, `${fixture.name}: ${errors.join("; ")}`);
  }
  const stored = fixtures.schemaCases.find((fixture) => fixture.definition === "CaseContract" && fixture.valid)?.value as CaseContract;
  assert.equal(stored.schemaVersion, 1);
});

test("canonical JSON and typed digest golden vectors", () => {
  for (const fixture of fixtures.canonicalCases) {
    assert.equal(canonicalize(fixture.value), fixture.canonical, fixture.name);
    assert.equal(typedDigest(fixture.domain, fixture.value), fixture.sha256, fixture.name);
  }
  for (const fixture of fixtures.canonicalRejectCases) {
    assert.throws(() => canonicalize(fixture.value), undefined, fixture.name);
  }
});

test("generated artifacts are marked and strict definitions reject unknown fields", async () => {
  const generated = await readFile(new URL("./generated/typescript/types.ts", import.meta.url), "utf8");
  assert.match(generated, /Code generated/);
  const base = fixtures.schemaCases.find((fixture) => fixture.definition === "CaseState" && fixture.valid)?.value as Record<string, unknown>;
  assert.ok(validator.validateDefinition("CaseState", { ...base, extra: true }).length > 0);
});

test("schema subset fails closed and protocol numbers are safe integers", () => {
  const invalidSchemas: unknown[] = [
    { $defs: { Probe: { type: "string", unsupportedKeyword: true } } },
    { $defs: { Probe: { $ref: "#/$defs/Target", type: "string" }, Target: { type: "string" } } },
    { $defs: { Probe: { oneOf: [{ type: "string" }], type: "string" } } },
    { $defs: { Probe: { type: "string" } }, allOf: [{ $ref: "#/$defs/Probe" }] },
    { $defs: { Probe: { type: "object", properties: {}, required: "x", additionalProperties: false } } },
    { $defs: { Probe: { type: "integer", minimum: "1" } } },
    { $defs: { Probe: { enum: "x" } } },
    { $defs: { Probe: { type: "integer", pattern: "x" } } },
    { $defs: { Probe: { type: "integer", format: "date-time" } } },
    { $defs: { Probe: { type: "string", format: "email" } } },
    { $defs: { Probe: { $ref: "#/$defs/Probe" } } },
    { $defs: { A: { $ref: "#/$defs/B" }, B: { $ref: "#/$defs/A" } } },
    { $defs: { Probe: { oneOf: [{ type: "string" }, { $ref: "#/$defs/Probe" }] } } },
  ];
  invalidSchemas.forEach((candidate) => assert.throws(() => new SchemaValidator(candidate as SchemaDocument)));

  const numberValidator = new SchemaValidator({
    $defs: { NumberProbe: { type: "number", minimum: 1, maximum: 3 } },
  } as SchemaDocument);
  assert.deepEqual(numberValidator.validateDefinition("NumberProbe", 2), []);
  assert.ok(numberValidator.validateDefinition("NumberProbe", 0).length > 0);
  assert.ok(numberValidator.validateDefinition("NumberProbe", 4).length > 0);
  assert.ok(numberValidator.validateDefinition("NumberProbe", 1.5).length > 0);
  assert.ok(numberValidator.validateDefinition("NumberProbe", Number.MAX_SAFE_INTEGER + 1).length > 0);
});
